### PR TITLE
[ticket/13162] Add truncate table functionality to DBAL

### PR DESCRIPTION
**Pull request by akbarhashmi**
PHPBB3-13162

Checklist:

- [x] Correct branch: 3.2.x for fixes
- [ ] Tests pass
- [ ] Code follows coding guidelines: [3.2.x]
- [x] Commit follows commit message 

Tracker ticket :

https://tracker.phpbb.com/browse/PHPBB3-13162

### CODE REVIEW

**Comment by DavidIQ** [comment]
`TRUNCATE TABLE` doesn’t work/exist in SQLite.

**Comment by akbarhashmi** [comment]
I see
Do I use the `DELETE FROM TABLE` command?
Or do I `DROP` the table and recreate it?

**Comment by 3D-I** [comment]
something alike
```
		switch ($this->db->get_sql_layer())
		{
			case 'sqlite':
			case 'sqlite3':
				$this->db->sql_query('DELETE FROM ' . $table_name);
			break;

			default:
				$this->db->sql_query('TRUNCATE TABLE ' . $table_name);
			break;
		}
```

**Comment by akbarhashmi** [comment]
Makes sense, I tried something similar

File: phpBB/phpbb/db/tools/tools.php [hunk 1]
```diff
@@ -1413,6 +1413,33 @@ function sql_table_drop($table_name)
...
+		$statements = array();
+
+		if (!$this->sql_table_exists($table_name))
+		{
+			return $this->_sql_run_sql($statements);
+		}
+
+		switch ($this->sql_layer)
+		{
+			case 'sqlite':
```
Comment (3D-I) [inline phpBB/phpbb/db/tools/tools.php]: I see there is only sqlite3 available here
https://github.com/akbarhashmi/phpbb/blob/5186570bac0af2a96ea8770c4436bb73b35c7763/phpBB/phpbb/db/tools/tools.php#L139

Comment (akbarhashmi) [reply phpBB/phpbb/db/tools/tools.php]: Removed case: sqlite 
Hope it passes this time

Comment (3D-I) [reply phpBB/phpbb/db/tools/tools.php]: Looks good, the Devs will say though.

**Comment by Nicofuma** [comment]
Please adds a test to ensure it works as expected on all supported databases.

**Comment by akbarhashmi** [comment]
```
public function test_table_truncate()
	{
		$this->tools->sql_create_table('prefix_test_table',
			array('COLUMNS' => array(
				'foo' => array('UINT', 42)))
		);

		$this->assertTrue($this->tools->sql_table_exists('prefix_test_table'));

		$this->tools->sql_table_truncate('prefix_test_table');

		$this->assertTrue($this->tools->sql_table_exists('prefix_test_table'));
	}
```
Will this take care of it? Or do we need to add such as function sql_table_empty ?

**Comment by Nicofuma** [comment]
you should insert a line before the truncate and then assert that the table is empty

File: phpBB/phpbb/db/tools/tools_interface.php [hunk 2]
```diff
@@ -23,6 +23,7 @@
 	 * Expected structure...
 	 * Key being one of the following
 	 *	drop_tables: Drop tables
+	 *  truncate_tables: Trancate tables (clear entire table safely)
```
Comment (3D-I) [inline phpBB/phpbb/db/tools/tools_interface.php]: Tr**a**ncate ----> Tr**u**ncate

**Comment by akbarhashmi** [comment]
Could someone figure out why this fails?

**Comment by DavidIQ** [comment]
According to this your test seems to be wrong:
https://travis-ci.org/phpbb/phpbb/jobs/346824143#L764

File: tests/dbal/db_tools_test.php [hunk 3]
```diff
@@ -344,6 +344,33 @@ public function test_table_drop()
...
+		$result = $this->db->sql_query($sql);
+		$this->assertTrue($this->db->sql_fetchrow($result));
+		$this->db->sql_freeresult($result);
+
+		$this->tools->sql_table_truncate('prefix_test_table');
+
+		// table empty after truncate
+		$sql = 'SELECT * FROM prefix_test_table';
+		$result = $this->db->sql_query($sql);
+		$this->assertTrue(! $this->db->sql_fetchrow($result));
```
Comment (CHItA) [inline tests/dbal/db_tools_test.php]: `$this->assertEquals(0, $this->db->get_row_count('prefix_test_table'));` would be a bit clearer (and it also might work)

**Comment by akbarhashmi** [comment]
appveyor fails with this for all db tests
Exception: Unable to connect to sqlsrv:Server=.\sql2016;Database=phpbb_test using PDO with error: could not find driver
What does this mean?

**Comment by CHItA** [comment]
I believe it has nothing to do with your code.

**Comment by akbarhashmi** [comment]
Can this be merged now that the checks have passed?

**Comment by 3D-I** [comment]
Why has been closed?

**Comment by akbarhashmi** [comment]
Accidently deleted this branch

**Comment by rubencm** [comment]
Closed in favor of #5329

File: phpBB/phpbb/db/tools/tools.php [hunk 4]
```diff
@@ -1413,6 +1413,36 @@ function sql_table_drop($table_name)
...
+	/**
+	 * {@inheritDoc}
+	 */
+	function sql_table_truncate($table_name)
+	{
+		$statements = array();
+
+		if (!$this->sql_table_exists($table_name))
+		{
+			return $this->_sql_run_sql($statements);
```
[the remark this answers is not in the document]
Comment (rubencm) [reply phpBB/phpbb/db/tools/tools.php]: sql_table_drop is programmed in that way.
If $this->return_statements is true, the method should return there an empty array, not a boolean

File: phpBB/phpbb/db/tools/tools.php [hunk 5]
```diff
@@ -1413,6 +1413,36 @@ function sql_table_drop($table_name)
...
+		}
+
+		switch ($this->sql_layer)
```
[the remark this answers is not in the document]
Comment (rubencm) [reply phpBB/phpbb/db/tools/tools.php]: A switch is cleaner i think

**Pull request closed without merging**